### PR TITLE
Mvc to mvp

### PR DESCRIPTION
**Pull request by rickramirr**
#### Code review checklist [Code Review Rubric](https://github.com/wizeline/iOS-apprenticeship-program-2020/blob/main/code-review-rubric.md)

1. DRY Principle (Don’t repeat yourself)

    - [ ] Excellent    
    - [ ] Good    
    - [ ] Adequate
    - [ ] Developing

2. Yagni Principle (You Aren't Gonna Need It)

    - [ ] Excellent    
    - [ ] Good    
    - [ ] Adequate
    - [ ] Developing

3. KISS Principle (Keep It Simple, St...d)

    - [ ] Excellent    
    - [ ] Good    
    - [ ] Adequate
    - [ ] Developing

4. Remove unused code

    - [ ] Excellent    
    - [ ] Good    
    - [ ] Adequate
    - [ ] Developing

5. Hard-coded values

    - [ ] Excellent    
    - [ ] Good    
    - [ ] Adequate
    - [ ] Developing

6. Language style guide & coding conventions
 
 - [Wizeline Style Guide](https://github.com/wizeline/mobile-documentation/blob/master/guidelines/swift-style-guide.md) 
 - [RayWenderlich Style Guide](https://github.com/raywenderlich/swift-style-guide)

    - [ ] Excellent    
    - [ ] Good    
    - [ ] Adequate
    - [ ] Developing

7. Architecture / Design Pattern

    - [ ] MVC  
    - [ ] MVP 
    - [ ] MVVM 
    - [ ] VIPER 

8. Architecture Implementation

    - [ ] Excellent    
    - [ ] Good    
    - [ ] Adequate
    - [ ] Developing

9. SOLID

- Single Responsibility Principle 

    - [ ] Excellent    
    - [ ] Good    
    - [ ] Adequate
    - [ ] Developing

- Open Closed Principle

    - [ ] Excellent    
    - [ ] Good    
    - [ ] Adequate
    - [ ] Developing

- Liskov substitutability principle

    - [ ] Excellent    
    - [ ] Good    
    - [ ] Adequate
    - [ ] Developing

- Interface segregation

    - [ ] Excellent    
    - [ ] Good    
    - [ ] Adequate
    - [ ] Developing

- Dependency Inversion principle

    - [ ] Excellent    
    - [ ] Good    
    - [ ] Adequate
    - [ ] Developing

10. Testing

    - [ ] Excellent    
    - [ ] Good    
    - [ ] Adequate
    - [ ] Developing

11. Force unwrapping !

    - [ ] Excellent    
    - [ ] Good    
    - [ ] Adequate
    - [ ] Developing

12. Mutable vs immutable property (Right use of `let`  and `var`)

    - [ ] Excellent    
    - [ ] Good    
    - [ ] Adequate
    - [ ] Developing

13. Retain cycles (Capture list)

    - [ ] Excellent    
    - [ ] Good    
    - [ ] Adequate
    - [ ] Developing

14. Code Documentation

    - [ ] Excellent    
    - [ ] Good    
    - [ ] Adequate
    - [ ] Developing

### CODE REVIEW

File: TheMovieDb/Modules/Home/HomeViewPresenter.swift [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+import Foundation
+import OSLog
+
+protocol HomeViewPresenterDelegate: AnyObject {
+    func didStartLoading()
+    func didFinishLoading()
+    func didStartSearching()
+    func didFinishSearching()
+    func didUpdateMovies(_ movies: [Movie])
+}
```
Comment (Carlos-Barron) [inline TheMovieDb/Modules/Home/HomeViewPresenter.swift]: Make a separate file

File: TheMovieDb/Modules/Home/HomeViewPresenter.swift [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+class HomeViewPresenter {
```
Comment (Carlos-Barron) [inline TheMovieDb/Modules/Home/HomeViewPresenter.swift]: final

File: TheMovieDb/Modules/Home/HomeViewPresenter.swift [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    
+    let service: MovieFeedRepository
```
Comment (Carlos-Barron) [inline TheMovieDb/Modules/Home/HomeViewPresenter.swift]: private

File: TheMovieDb/Modules/Home/HomeViewPresenter.swift [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    }
+    
+    var currentFeed: FeedTypes = .trending {
+        didSet {
+            resetFeed()
+            if !isSearching {
+                getMoviesIfNeeded(search: nil)
+            }
+        }
+    }
```
Comment (Carlos-Barron) [inline TheMovieDb/Modules/Home/HomeViewPresenter.swift]: The view controller can't be allowed to access the Models

File: TheMovieDb/Modules/Home/HomeViewPresenter.swift [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    
+    func getMoviesIfNeeded(search: String? = nil) {
+        guard !isLoading else {
+            return
+        }
+        isLoading = true
+        service.getMovieFeed(
+            on: currentFeed,
+            page: loadedPages + 1,
```
Comment (Carlos-Barron) [inline TheMovieDb/Modules/Home/HomeViewPresenter.swift]: Magic number

File: TheMovieDb/Modules/Home/HomeViewPresenter.swift [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+            query: search
+        ) { [weak self] result in
+            switch result {
+            case .success(let response):
+                self?.loadedPages = response.page
+                self?.movies.append(contentsOf: response.results)
+                guard let movies = self?.movies else {
+                    return
+                }
+                self?.delegate?.didUpdateMovies(movies)
```
Comment (Carlos-Barron) [inline TheMovieDb/Modules/Home/HomeViewPresenter.swift]: Refactor to a unique function

File: TheMovieDb/Modules/Home/HomeViewPresenter.swift (hunk 1 above)
Comment (rickramirr) [reply TheMovieDb/Modules/Home/HomeViewPresenter.swift]: Solved

File: TheMovieDb/Modules/Home/HomeViewPresenter.swift (hunk 3 above)
Comment (rickramirr) [reply TheMovieDb/Modules/Home/HomeViewPresenter.swift]: Solved

File: TheMovieDb/Modules/Home/HomeViewPresenter.swift (hunk 4 above)
Comment (rickramirr) [reply TheMovieDb/Modules/Home/HomeViewPresenter.swift]: Solved

File: TheMovieDb/Modules/Home/HomeViewPresenter.swift (hunk 5 above)
Comment (rickramirr) [reply TheMovieDb/Modules/Home/HomeViewPresenter.swift]: I tried to solve this creating a new property

File: TheMovieDb/Modules/Home/HomeViewPresenter.swift (hunk 6 above)
Comment (rickramirr) [reply TheMovieDb/Modules/Home/HomeViewPresenter.swift]: solved

**Review by Carlos-Barron — APPROVED** [review]

**Pull request merged**